scipy/optimize/_differentialevolution.py

```diff
@@ -364,13 +364,16 @@ def __init__(self, func, bounds, args=(),
         self.__scale_arg1 = 0.5 * (self.limits[0] + self.limits[1])
         self.__scale_arg2 = np.fabs(self.limits[0] - self.limits[1])
 
-        parameter_count = np.size(self.limits, 1)
+        self.parameter_count = np.size(self.limits, 1)
+
         self.random_number_generator = _make_random_gen(seed)
 
-        #default initialization is a latin hypercube design, but there
-        #are other population initializations possible.
-        self.population = np.zeros((popsize * parameter_count,
-                                    parameter_count))
+        # default population initialization is a latin hypercube design, but
+        # there are other population initializations possible.
+        self.population = np.zeros((popsize * self.parameter_count,
+                                    self.parameter_count))
```

Comment:
There was a reason for having the current scheme. The `solve` method takes the *current* state of the population and evolves it by `maxiter` generations. One can check how the population evolves over time by setting `maxiter` to an arbitrary number (e.g. 1), and calling `solve` repeatedly, each time checking what the population is doing.  For this reason you wouldn't want to initialise the population before each call to solve.
A second, equally important reason, is that there are restrictions on the population shape (number of columns needs to be same length as `bounds`).

A future change to this class could be to extract the `for nit in range(1, self.maxiter + 1):` loop into a separate method, and make that method into a generator. Each call to the generator would evolve the population by a single generation.